From a31bb3f1b4d03706a56df128c3ecd5c9c4ed4b55 Mon Sep 17 00:00:00 2001
Message-Id: <a31bb3f1b4d03706a56df128c3ecd5c9c4ed4b55.1440106296.git.chang-joon.lee@intel.com>
In-Reply-To: <1f426693c14c79e65ad8e62ef6bb4b07d6484040.1440106296.git.chang-joon.lee@intel.com>
References: <1f426693c14c79e65ad8e62ef6bb4b07d6484040.1440106296.git.chang-joon.lee@intel.com>
From: Arun R Murthy <arun.r.murthy@intel.com>
Date: Tue, 11 Aug 2015 14:29:54 +0530
Subject: [PATCH 11/11] SQUASHME! [VPG]: calculate the aspect ratio for pfit

When a pfit enable request comes, calculate the aspect ratio,
determine the mode in which pfit can be enabled and verify with
the requested mode, if not same use the calculated mode to enable
panel fitter.
panel resolution 1920 x 1080, aspect ratio = 1920/1080 = 1.77
image to be scaled 1024 x 768, aspect ratio = 1024 / 756 = 1.33
    1920 / h = 1024 / 768
    h = 1440, which is greater then the panel height
    w / 1080 = 1024 / 768
    w = 1440
    hence panel fitter can be enabled in pillar box mode.

SQUASHME: This patch is to be squashed with the below patch
    commit 32ebc82f175e8ddad7a2a2d6aa0b7b13fa9cb15d
    Author: Arun R Murthy <arun.r.murthy@intel.com>
    Date:   Tue Mar 3 14:27:59 2015 +0530

    SQUASHME! [VPG]: drm/i915: Enable panel fitter only on HWC request

Change-Id: Ib233536cb0fc89521efd976e0fa7556d2725b679
Tracked-On: https://vthsd.fm.intel.com/hsd/pcgsw/default.aspx#bug/default.aspx?bug_id=8158848
Signed-off-by: Arun R Murthy <arun.r.murthy@intel.com>
---
 drivers/gpu/drm/i915/intel_display.c | 19 +++++++++++++++++++
 1 file changed, 19 insertions(+)

diff --git a/drivers/gpu/drm/i915/intel_display.c b/drivers/gpu/drm/i915/intel_display.c
index 5ea7005..44e1918 100644
--- a/drivers/gpu/drm/i915/intel_display.c
+++ b/drivers/gpu/drm/i915/intel_display.c
@@ -11113,6 +11113,8 @@ int intel_set_disp_calc_flip(struct drm_mode_set_display *disp,
 							PFIT_PIPE_SHIFT));
 			DRM_ERROR("Not enabling Panel Fitter\n");
 		} else {
+			u32 pfit_mode = 0;
+
 			pfit_control &= ~PFIT_PIPE_MASK;
 			pfit_control |= (intel_crtc->pipe << PFIT_PIPE_SHIFT);
 
@@ -11120,6 +11122,16 @@ int intel_set_disp_calc_flip(struct drm_mode_set_display *disp,
 				(((disp->panel_fitter.src_w - 1) << 16) |
 						(disp->panel_fitter.src_h - 1));
 
+			if (((mode->hdisplay * disp->panel_fitter.src_h) /
+						disp->panel_fitter.src_w) < mode->vdisplay)
+				pfit_mode |= PFIT_SCALING_LETTER;
+			else if (((mode->vdisplay * disp->panel_fitter.src_w) /
+						disp->panel_fitter.src_h) < mode->hdisplay)
+				pfit_mode |= PFIT_SCALING_PILLAR;
+			else if ((mode->hdisplay / mode->vdisplay) ==
+					(disp->panel_fitter.src_w / disp->panel_fitter.src_h))
+				pfit_mode |= PFIT_SCALING_AUTO;
+
 			/* Enable Panel fitter if any valid mode is set */
 			pfit_control = (1 << 31) | pfit_control;
 			if (disp->panel_fitter.mode == AUTOSCALE) {
@@ -11139,6 +11151,13 @@ int intel_set_disp_calc_flip(struct drm_mode_set_display *disp,
 						(((mode->hdisplay - 1) << 16) |
 							(mode->vdisplay - 1));
 			}
+
+			if (pfit_mode != (pfit_control & ~MASK_PFIT_SCALING_MODE)) {
+				DRM_ERROR("pfit mode(%d) mismatch with calculated mode(%d)\n",
+						(pfit_control & ~MASK_PFIT_SCALING_MODE), pfit_mode);
+				pfit_control &=  MASK_PFIT_SCALING_MODE;
+				pfit_control |= pfit_mode;
+			}
 			intel_crtc->pfit_control = pfit_control;
 			if (pfit_control != pfitcontrol)
 				dev_priv->pfit_changed = true;
-- 
1.9.1

